src/main/java/org/elasticsearch/script/mustache/MustacheScriptEngineService.java

```diff
@@ -98,22 +99,26 @@ public Object compile(String template) {
      * @return the processed string with all given variables substitued.
      * */
     @Override
-    public Object execute(Object template, Map<String, Object> vars) {
-        BytesStreamOutput result = new BytesStreamOutput();
-        UTF8StreamWriter writer = utf8StreamWriter().setOutput(result);
-        ((Mustache) template).execute(writer, vars);
+    public Object execute(CompiledScript template, Map<String, Object> vars) {
         try {
-            writer.flush();
-        } catch (IOException e) {
-            logger.error("Could not execute query template (failed to flush writer): ", e);
-        } finally {
+            BytesStreamOutput result = new BytesStreamOutput();
+            UTF8StreamWriter writer = utf8StreamWriter().setOutput(result);
+            ((Mustache) template.compiled()).execute(writer, vars);
             try {
-                writer.close();
+                writer.flush();
             } catch (IOException e) {
-                logger.error("Could not execute query template (failed to close writer): ", e);
+                logger.error("Could not execute query template [" + template.name() + "] (failed to flush writer): ", e);
+            } finally {
+                try {
+                    writer.close();
+                } catch (IOException e) {
+                    logger.error("Could not execute query template [" + template.name() + "] (failed to close writer): ", e);
+                }
             }
+            return result.bytes();
+        } catch (Exception e) {
+            throw new ScriptException("Error with query template [" + template.name() + "]: " + e.getMessage(), e);
         }
```

Comment:
Can we simplify this to use try-with-resources? close() should call flush() anyway, and then this way no exceptions are ever hidden, instead they are chained with suppression: e.g.

    try (UTF8StreamWriter writer = ...) {
      execute();
    } catch (Exception e) {
      logger.error(xxx);
      throw new ScriptException(....)
    }